### ISSUE TITLE
Improve documentation (and examples) of Blocking(re)Connection.

### ISSUE THREAD

**Issue by Darkless012** [pika/pika#757]
I will start with 2 basic questions which I will accompany with our use-case example.

1) How to have long running connection on BlockingConnection?
(and answer: "Use another Adapter" is not an answer at all - see the use-case)

2) How to properly react on disconnect while running BlockingConnection
(or let me rephrase it: How to do reconnection on BlockingConnection properly?)

Our use-case:
I am enforced to use Blocking Connection because of possibility to "get" message when I need the message. (Using Apache Storm - It will pick message from Queue and process it somehow) And I didn't find a way to do it in asynchronous manner.

The thing in Apache Storm that picks messages from Rabbit is called Spout. And Spout should be running indefinitely to pick messages when IT needs to pick.
There is loop called that will pick the message, does something with it and emit to another processing part - So I don't have an option to run `start_consuming()` -> Thus need for Blocking Connection (or I am missing example/documentation how to do it in another, async, fashion)

Since it needs to run indefinitely, it needs to read messages "forever". But Blocking Connection will close the connection after a while, since it does not implement "hearthbeat_checker" directly.

Ok, how do I solve the connection issues:
Try to run hearthbeat checker? (Disabling hearthbeats on server is not an option). Which leads to using code:

```
# Run hearthbeat
self.connection._impl._create_heartbeat_checker()
```

But is this OK? I don't know. There is literally NONE in the documentation, so it is my wild guess - based on reading the code.
But I suppose it will not call the hearthbeats itself, because there were a mention in docs about timer not being processed - need to call another method.
Ok, so I'm calling `process_data_events()` before every `connection.basic_get()`. Which should be often enough. But will it be enough?

And sometime I still got disconnected. And the only time I know connection was closed is when `basic_get()` (and `basic_publish()`) throws an Exception. (Even `connection.is_open()` is returning always True, although connection is already closed.)

Which leads to implementation of methods called like `retry_basic_get()` and `retry_basic_publish()` That tries to get/publish message and on error will make disconnect/connect to server and tries to do the getting/publishing again.

This can ensure that the message is obtained / published in fairly stable manner. 
But it generates lots of errors like `Read empty data, calling disconnect` or `Connection close detected` or even `Fatal Socket Error: error(32, 'Broken pipe')`

So these errors will stack in the LOG, but I have literally no option to react on them, since Blocking Connection has no "on_disconnect" method. Or disable them, since they are not Errors anymore - because I'm trying to close connection for good and then open a new one.

Am I terribly missing something ?
Or could be Blocking Connection better documented?
Or could be Blocking Connection better designed?
Or could async Adapter be used for this? (I've tried without success)

I feel that it is lot of fuzz to use Blocking Connection in Pythonic manner for any purposes than to just get/publish few messages and then to close the connection.

**Comment by antoineleclair** [comment]
Same problem here. I'm not sure if it's a bug or if I'm missing something.

I ended up with something like this to reconnect on __any__ error.


```python
class Producer:
    """Message producer.

    This class is used to send message to RabbitMQ.

    """
    def __init__(self):
        log.debug('Initializting MQ producer')
        self._connect()

    def _connect(self):
        params = pika.URLParameters(config.get('rabbitmq.url'))
        self.connection = pika.BlockingConnection(params)
        self.channel = self.connection.channel()
        common.declare(self.channel)

    def publish(self, queue, msg):
        """Publishes a message to RabbitMQ.

        This is not thread safe.

        """
        log.debug('Publishing message to queue %s: %s', queue, msg)
        try:
            self._publish(queue, msg)
        except:
            log.info('Got exception when trying to publish MQ message, '
                     'reconnecting')
            self._connect()
            self._publish(queue, msg)

    def _publish(self, queue, msg):
        self.channel.basic_publish(exchange='',
                                   routing_key=queue,
                                   body=msg,
                                   properties=pika.BasicProperties(
                                      delivery_mode = 2, # persistent
                                   ))

    def close(self):
        try:
            self.connection.close()
        except pika.exceptions.ConnectionClosed:
            pass

```
I don't mind reconnecting on errors, but it would be great to not have to silent all errors like this. I'm not sure if I missed something (because the doc is unclear) or if it's an actual issue.

**Comment by vitaly-krugl** [comment]
The pull request #956 might be helpful with the new method `add_callback_threadsafe`.

In general, you might need to wrap the connection and use a decorator for reconnecting.

For concept, see [`_RETRY_ON_AMQP_ERROR` decorator and `_ChannelManager` transport wrapper class](https://github.com/numenta/numenta-apps/blob/46fff2aec40c1a4d122f575ec0e3edeac0456068/nta.utils/nta/utils/message_bus_connector.py). This specific implementation is now using the `haigha` client under the covers (for licensing reasons), but used to use pika BlockingConnection in the same way.

**Comment by vitaly-krugl** [comment]
This should be addressed by pull request #1028. Closing...
